<!--961032830987546d0e6d54829fc886f6-->

目录(Catalo)

* [异常捕捉](#%E5%BC%82%E5%B8%B8%E6%8D%95%E6%8D%89)
  * [主动抛出异常](#%E4%B8%BB%E5%8A%A8%E6%8A%9B%E5%87%BA%E5%BC%82%E5%B8%B8)

<!--a46263f7a69f33f39fc26f907cdb773a-->
# 异常捕捉

```python
try:
    # 检测异常的代码
    pass
except ValueError:
    print("数值错误")
except TypeError:
    print("类型错误")
except Exception as result:
    print("未知错误 %s " % result)
else:
    # 没有异常才会被执行的代码
    pass
finally:
    # 无论是否发生异常都会被执行的代码
    pass
```

**如果函数或方法发生异常时，程序并不会立即终止，而是会将异常传递给调用函数或方法的一方，如果在主程序中异常还是没得到处理，程序就会终止**

* Exception 是python中专门针对异常的类

## 主动抛出异常

主动抛出异常需要先实例化一个Exception 对象，然后使用`raise`抛出异常，然后可以捕捉这个异常

```python
def input_psw():
    psw=input("请输入密码： ")
    if len(psw)<8:
        error=Exception("密码长度不够")
        raise error
    else:
        return psw
try:
    s=input_psw()
except Exception as error:
    print(error)
else:
    print(s)
finally:
    print("---end---")
```

```txt
请输入密码： 1
密码长度不够
---end---
```